### ISSUE TITLE
Function to calculate number of tokens

### ISSUE THREAD

**Issue by rmilkowski** [openai/openai-python#412]
### Describe the feature or improvement you're requesting

It would be useful if the module provided a function to calculate number of token in a given prompt for a given model, without having to use another 3rd party modules to do so.

One examle when it would be useful is to trim fed history if the entire prompt (+max_tokens) is above a given model limit, before sending the query.

**Comment by lewiswilliam** [comment]
Hi @rmilkowski 

If you are using the python sdk, tiktoken is the library created by OpenAI itself to tokenise/count the tokens:
https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb

Also, check the following link which describes how to use the token count:
https://github.com/openai/openai-cookbook/blob/main/examples/Question_answering_using_embeddings.ipynb

Please check if this helps.

**Comment by rmilkowski** [comment]
fair enough, although it's a shame they provide an example code how to handle different chat models instead of providing a function to deal with it, so users wouldn't have to keep their code updated and would have pick up new models by simply upgrading the module. I submitted https://github.com/openai/tiktoken/issues/115

**Comment by StephenHodgson** [comment]
I agree, even an api call that calculates tokens based on model and text input would be super valuable

**Comment by jussker** [comment]
This is a great solution and it's useful. 

https://gist.github.com/CGamesPlay/dd4f108f27e2eec145eedf5c717318f5

**Comment by StephenHodgson** [comment]
Personally, I believe this should actually be done server side via API endpoint.

**Comment by ldorigo** [comment]
If OpenAI isn't planning on providing this functionality, could you _at least_ provide some clear informations on how tokens are counted for function calls/tool calls/structured outputs/etc. etc. etc. ? How are we supposed to calculate token usage?

**Comment by Ai80-8** [comment]
Stop coping

**Comment by MukundaKatta** [comment]
Built-in token counter is the right ask. Three reasons it's never landed: tiktoken is the de facto answer but doesn't track newer additions (reasoning, audio, image tokens) consistently, model-id-to-encoder mapping drifts every model release, and tool-schema and image inputs need separate counting logic.

A small client.tokens.count(model, messages, tools=None, images=None) -> {input_tokens, breakdown} call would be the right primitive. Server-side estimate avoids the maintenance burden the client-side tiktoken approach has.

Until then, count_tokens on Anthropic and Bedrock's CountTokens at least give the rest of the ecosystem a consistent shape to copy.

**Comment by west0nG** [comment]
This is still a gap worth filling. In the meantime, the practical path is using tiktoken directly for pre-flight token counting — OpenAI maintains it and it's designed for exactly this use case.

**For history trimming specifically**, here's a working pattern:

```python
import tiktoken

def count_tokens_for_messages(messages: list[dict], model: str = "gpt-4o") -> int:
    enc = tiktoken.encoding_for_model(model)
    tokens = 0
    for msg in messages:
        tokens += 4  # role + content framing overhead per message
        for key, value in msg.items():
            if isinstance(value, str):
                tokens += len(enc.encode(value))
            elif isinstance(value, list):
                # multimodal content blocks
                for block in value:
                    if isinstance(block, dict) and block.get("type") == "text":
                        tokens += len(enc.encode(block["text"]))
    tokens += 2  # reply priming
    return tokens

def trim_history(messages, model, max_tokens, max_completion_tokens):
    budget = max_tokens - max_completion_tokens
    while count_tokens_for_messages(messages, model) > budget and len(messages) > 1:
        # Remove oldest non-system message
        messages = [messages[0]] + messages[2:] if messages[0]["role"] == "system" else messages[1:]
    return messages
```

Note: tool call messages add framing tokens beyond what string encoding captures, so leave a ~5-10% buffer for safety.

OpenAI also now exposes a [token counting endpoint](https://developers.openai.com/api/reference/resources/responses/subresources/input_tokens/methods/count) for the Responses API that returns the exact server-side count — the most accurate option for preflight checks.

For quick interactive token budgeting without code (e.g., testing whether a document fits in a context window), https://llmtokens.vercel.app shows per-model token counts and cost estimates for pasted text.

**Comment by marcuswood-oai** [comment]
Thank you for the request and discussion. The SDK now provides `client.responses.input_tokens.count(model=..., input=..., tools=...)`, available since v2.6.0. It counts the request before generation, including message framing and tool definitions; see the [token-counting guide](https://developers.openai.com/api/docs/guides/token-counting).

Please pass the same inputs and configuration you intend to send to Responses, then use the returned `input_tokens` to budget or trim your input. This makes a server request and uses Responses-format inputs; existing Chat Completions integrations need to adapt their payloads. We’re closing because this capability is now available.
